.. _dataformat:

===============
Data format
===============
id
hicstuff can generate sparse Hi-C contact maps either in `instagraal <https://github.com/koszullab/instaGRAAL>`_ compatible coordinate format or in `cooler <https://github.com/mirnylab/cooler>`_ -compatible 2D bedgraph format. 

**COOrdinate matrix**

This is the default format for sparse matrices generated by hicstuff pipeline. This is a tab-separated file with 3 columns: 
frag1, frag2, contacts. The frag columns correspond to the absolute id of the restriction fragments (0-indexed). The first row is a header containing the number of rows, number of columns and number of nonzero entries in the matrix.

+-------+-------+----------+
| frag1 | frag2 | contacts |
+-------+-------+----------+ 
 

 

**Bedgraph 2D**

Sparse matrices can be generated in this format optionally by using the flag :code:`-M cooler`. It has two fragments/bins per line, and the number of times they are found together. It is a tab-separated format with the following fields: 

+--------+--------+------+--------+--------+------+-------+
| chrom1 | start1 | end1 | chrom2 | start2 | end2 | value | 
+--------+--------+------+--------+--------+------+-------+

Those files can be loaded by cooler using :code:`cooler load -f bg2 <chrom.sizes>:<binsize> in.bg2.gz out.cool` where chrom.sizes is a tab delimited file with chromosome names and length on each line, and binsize is the size of bins in the matrix.

**Pairs format**

All intermediate files are stored in the pairs format, `as defined by the 4DN consortium <https://github.com/4dn-dcic/pairix/blob/master/pairs_format_specification.md>`_. Pairs file have 7 columns before fragment atribution:

+--------+------+------+------+------+---------+---------+
| readID | chr1 | pos1 | chr2 | pos2 | strand1 | strand2 | 
+--------+------+------+------+------+---------+---------+

and 9 columns afterwards:

+--------+------+------+------+------+---------+---------+-------+-------+
| readID | chr1 | pos1 | chr2 | pos2 | strand1 | strand2 | frag1 | frag2 | 
+--------+------+------+------+------+---------+---------+-------+-------+

Those files can be loaded by cooler using :code:`cooler cload pairs -c1 2 -p1 3 -c2 4 -p2 5 -f bg2 <chrom.sizes>:<binsize> in.pairs2.gz out.cool` where chrom.sizes is a tab delimited file with chromosome names and length on each line, and binsize is the size of bins in the matrix.


**fragments_list.txt**

This tab separated file provides information about restriction fragments positions, size and GC content. Note the coordinates are 0 point basepairs, unlike the pairs format, which has 1 point basepairs. 

* id: 1 based restriction fragment index within chromosome.
* chrom: Chromosome identifier. Order should be the same as in info_contigs.txt or pairs files.
* start_pos: 0-based start of fragment, in base pairs.
* end_pos: 0-based end of fragment, in base pairs.
* size: Size of fragment, in base pairs.
* gc_content: Proportion of G and C nucleotide in the fragment.

**info_contigs.txt**

This tab separated file gives information on contigs, such as number of restriction fragments and size. It has the following headers and columns:

* contig: Chromosome identified. Order should be the same in pairs files or fragments_list.txt.
* length: Chromosome length, in base pairs.
* n_frags: Number of restriction fragments in chromosome.
* cumul_length: Cumulative length of previous chromosome, in base pairs.

