{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# hicstuff command line interface demo\n",
    "\n",
    "## Getting started\n",
    "\n",
    "The easiest way to generate matrices is to use `hicstuff pipeline`. By default the command only requires two fastq files (forward and reverse reads) and a genome in fasta format.\n",
    "\n",
    "The pipeline command can be used to generate the Hi-C contact map from the input reads.\n",
    "\n",
    "```bash\n",
    "hicstuff pipeline --genome genome.fa \\\n",
    "                  --outdir results \\\n",
    "                  forward.fq \\\n",
    "                  reverse.fq\n",
    "```\n",
    "\n",
    "For instance, this will create a directory named \"results\", containing three output text files with tab-separated columns.\n",
    "\n",
    " * `abs_fragments_contacts_weighter.txt`: Sparse matrix file with 3 columns the rows, column and values of nonzero pixels. The first row contains the shape and total number of nonzero pixels in the matrix.\n",
    " * `fragments_list.txt`: Contains genomic coordinates of the matrix bins (row/columns).\n",
    " * `info_contigs.txt`: Contains chromosome names, theirs length and number of bins.\n",
    "\n",
    "### Using an indexed genome\n",
    "When given a genome in fasta format, hicstuff regenerates the index everytime unless it finds an index starting by the genome filename. For example using bowtie2, you can index the genome like this:\n",
    "\n",
    "```bash\n",
    "bowtie2-build genome.fa genome.fa\n",
    "```\n",
    "\n",
    "And when running hicstuff with `--genome=genome.fa`, it will automatically find the index files and not regenerate it.\n",
    "\n",
    "### Setting the binning\n",
    "\n",
    "By default, matrices generated are binned at 5kb. You can change this using the `--enzyme` option. This option allows to specify either a bin size, or an enzyme. For example if you set `--enzyme='DpnII'`, the matrix will be at restriction fragments resolution.\n",
    "\n",
    "### Temporary files\n",
    "\n",
    "By default temporary files are removed when the pipeline finishes. They can be kept by adding the `--no-cleanup` flag. For example, if you run:\n",
    "\n",
    "```bash\n",
    "hicstuff pipeline -g genome.fa --no-cleanup --prefix demo\n",
    "```\n",
    "The `--prefix` option used here gives a common prefix to all output files, overriding the default output file names as follows:\n",
    "\n",
    " * `abs_fragments_contacts_weighted.txt` -> `demo.mat.tsv`\n",
    " * `fragments_list.txt` -> `demo.frags.tsv`\n",
    " * `info_contigs.txt` -> `demo.chr.tsv`\n",
    " \n",
    "And the output folder should now contain a `tmp/` subfolder and look like this:\n",
    "```\n",
    "output\n",
    "├── demo.chr.tsv\n",
    "├── demo.frags.tsv\n",
    "├── demo.hicstuff_20190423185220.log\n",
    "├── demo.mat.tsv\n",
    "├── demo.distance_law.txt\n",
    "├── plots\n",
    "│   ├── event_distance.pdf\n",
    "│   ├── event_distribution.pdf\n",
    "│   └── frags_hist.pdf\n",
    "└── tmp\n",
    "    ├── demo.for.bam\n",
    "    ├── demo.genome.fasta\n",
    "    ├── demo.rev.bam\n",
    "    ├── demo.valid_idx_filtered.pairs\n",
    "    ├── demo.valid_idx.pairs\n",
    "    └── demo.valid.pairs\n",
    "```\n",
    "\n",
    "### Additional options\n",
    "\n",
    "The `hicstuff pipeline` command has additional options allowing to tweak various parameters and change the output files or their format. You can always consult `hicstuff pipeline --help` to get a comprehensive description of those options, but here are a few of them:\n",
    "\n",
    " * `--filter`: Filters out spurious 3C events, such as self religations or undigested fragments. This is only really useful at very fine resolutions (1-2kb) and not needed most of the time. This option is only meaningful when `--enzyme` is given a restriction enzyme and not a bin size.\n",
    " * `--duplicates`: Removes PCR duplicates, defined as sets of pairs having identical mapping positions for both reads.\n",
    " * `--matfmt`: allows to specify what file format should be used for the matrix. Available formats are bg2 (bedgraph2d), graal (the text format described above) and cool, [a binary format](https://cooler.readthedocs.io) that is probably the most appropriate for large genomes.\n",
    " * `--distance-law`: Computes the distance-law table (i.e. the probability of contacts as a function of genomic distance).\n",
    " * `--plot`: Enables plotting. When used in conjunction with `--filter` or `--distance-law`, this will generate figures showing properties of your Hi-C data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Advanced usage\n",
    "\n",
    "### Starting from intermediate files\n",
    "\n",
    "If your hicstuff run was interrupted, or you wish to align the reads separately, the `--start-stage` option allows the `hicstuff pipeline` command to take intermediate files as input. For example to skip the alignment step and start from aligned reads:\n",
    "\n",
    "```bash\n",
    "hicstuff pipeline --start-stage bam --genome genome.fa forward.bam reverse.bam\n",
    "```\n",
    "\n",
    "Or if you already have a pairs file:\n",
    "\n",
    "```bash\n",
    "hicstuff pipeline --start-stage pairs --genome genome.fa valid.pairs\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating the distance law\n",
    "\n",
    "The distance law is the probability of contact of two fragments in function of the distance between these fragments. There are two ways to compute it with hicstuff. The first one using the full pipeline with the option `--distance-law`, as done above. It's possible to add an option `--centromeres` if you want to compute the distance law on separate arms. The output of this command will be a raw table of the distance without any treatment of the data. It will be then possible with the command distancelaw to process this table.\n",
    "\n",
    "The second way is to use the command distancelaw with the pairs file as input:\n",
    "\n",
    "```bash\n",
    "hicstuff distancelaw --average \\\n",
    "                     --big-arm-only \\\n",
    "                     --centromeres centromeres.txt \\\n",
    "                     --frags output/demo.frags.tsv \\\n",
    "                     --inf 3000 \\ \n",
    "                     --outputfile-img output/demo_distance_law.svg \\ \n",
    "                     --labels labels.txt \\\n",
    "                     --sup 500000 \\\n",
    "                     --pairs output/tmp/demo.valid_idx_filtered.pairs\n",
    "```\n",
    "\n",
    "For instance, this will create an image with the distance law generated from the pairs file given in input. The distance law will be the average between all the distance laws of the arms bigger than 500kb. The logspace used to plot it will have a base 1.1 by default. The limits of the x axis will be 3kb and 500kb.\n",
    "\n",
    "Note that if `hicstuff pipeline` was given the `--distance-law` option, the output folder should contain a file named `distance_law.txt` containing the precomputed interaction frequencies. This file can be provided to the `hicstuff distancelaw` command using `--dist-tbl distance_law.txt` instead of the pairs file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Operations on Hi-C matrices\n",
    "\n",
    "All commands described below can take the output of hicstuff as input. They will accept either a bg2 matrix, a cool matrix or a graal matrix. Note that when using a graal matrix, you will usually need to specify the fragments file using `--frags` since genomic coordinates are not encoded in this matrix format.\n",
    "\n",
    "### Visualizing the matrix\n",
    "\n",
    "Below are example commands that can be used to visualise Hi-C matrices with hicstuff.\n",
    "\n",
    "```bash\n",
    "# Viewing a normalized (=balanced) matrix in graal format at 5kb resolution\n",
    "hicstuff view --binning 5kb --normalize --frags output/demo.frags.tsv output/demo.mat.tsv\n",
    "# Viewing a log ratio of 2 matrices in cool format\n",
    "hicstuff view sample1.cool sample2.cool\n",
    "# Viewing a raw matrix in bedgraph2 format at 500bp resolution with log transformed contacts\n",
    "hicstuff view --binning 500bp --transform log high_res.bg2\n",
    "```\n",
    "This will show an interactive heatmap using matplotlib. In order to save the matrix to a file instead, one could add `--output output/demo.png`\n",
    "\n",
    "Note there are others options allowing to process the matrix which are documented in the help message.\n",
    "\n",
    "### Converting between formats\n",
    "\n",
    "Output files from `hicstuff pipeline` can be converted between different format using the command `hicstuff convert`. For example to generate the file `cool_output/demo.cool` from files in the default hicstuff format:\n",
    "\n",
    "```bash\n",
    "hicstuff convert --frags output/demo.frags.tsv \\\n",
    "                 --chroms output/demo.chr.tsv \\\n",
    "                 --to cool \\\n",
    "                 output/demo.mat.tsv\n",
    "                 output/demo\n",
    "```\n",
    "\n",
    "Notice that the command takes 2 positional arguments, the first one is the matrix file and the second is the prefix to give for output files, to which the extension will be added depending on the output format chosen. Input format is inferred automatically from the input matrix.\n",
    "\n",
    "### Rebinning existing matrices\n",
    "\n",
    "Files previously produced by hicstuff pipeline can be rebinned at a lower resolutions using the `hicstuff rebin` command.\n",
    "This will generate a new matrix, a new fragments_list.txt and a new info_contigs.txt, all with updated number of bins:\n",
    "\n",
    "```bash\n",
    "hicstuff rebin -f output/demo.frags.tsv \\\n",
    "               -c output/demo.chr.tsv \\\n",
    "               --out rebin_1kb \\\n",
    "               --binning 1kb output/demo.mat.tsv\n",
    "```\n",
    "\n",
    "When working with cool or bedgraph2 files, the command is simpler as we don't need fragments and contig files:\n",
    "\n",
    "```bash\n",
    "# Rebins demo.cool at 10kb and saves the results to rebinned.cool\n",
    "hicstuff rebin --binning 10kb demo.cool rebinned\n",
    "```\n",
    "\n",
    "### Subsampling contacts\n",
    "\n",
    "For many applications, differences in sequencing coverage will impact results. To avoid this, one can subsample contacts from Hi-C matrices to ensure the different samples to be compared have comparable signal. This functionality is implemented in the `hicstuff subsample` command, which allows to keep a fixed number of contacts from a matrix, or to extract a fraction of contacts:\n",
    "\n",
    "```bash\n",
    "# Keep 30% contacts in matrix.cool and save the results to subsamples_30.cool\n",
    "hicstuff subsample --prop 0.5 matrix.cool subsample_30\n",
    "# Keep 1 million contacts in matrix.bg2 and save the result in subsample_1M.bg2\n",
    "hicstuff subsample --prop 1000000 matrix.bg2 subsample_1M\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
